### PR TITLE
Add module for aerc

### PR DESCRIPTION
**Pull request by lukasngl**
### Description

Module for the [aerc email client][aerc].

[aerc]: https://aerc-mail.org/

Related
  - #2894

Features
  - General configuration (cf. aerc-config(5)), via `aerc.extraConfig`
  - Configure local and global key bindings (cf. aerc-config(5)), via `aerc.extraBinds`
  - Configure theme with stylesets (cf. aerc-stylesets(7)), via `aerc.stylesets`
  - Create email templates (cf. aerc-templates(7)), via `aerc.templates`
  - Manual account configuration (cf. aerc-config(5)), via `aerc.extraAccounts`
  - Automagically setup smtp, imap, msmtp and maildir (via offlineimap or mbsync), using `email.accounts.<name>`
  - Populate `accounts.conf` with defaults from `email.accounts.<name>`. E.g., `realName`, `address` and `folders`
  - Contextual account configuration (cf. aerc-config(5)), via `mail.accounts.<name>.extra{Accounts,Binds,Config}`

Non-Features:
  - Write a query-map for notmuch (cf. aerc-notmuch(5)).
    If needed, one can just use `pkgs.writeText`, e.g.:
    ```nix
    accounts.email.<name>.aerc.extraConfig.query-map =
      pkgs.writeText "querymap" "inbox=tag:inbox and not tag:archived";
    ```
  - Setup maildir if neither offlineimap nor mbsync are enabled.
    The value `email.accounts.<name>.maildir` is not null by default. So there is no way of knowing if the user wants a setup with maildir, except if offlineimap or mbsync are enabled.
    If needed, one could just manually set it up like this:
    ```nix
    accounts.email.<name>.aerc.extraConfig.source =
      "maildir://${config.accounts.mailDirBasePath}/${config.accounts.<name>.maildir}"
    ```

### Checklist

- [X] Change is backwards compatible.

- [X] Code formatted with `./format`.

- [X] Code tested through `nix-shell --pure tests -A run.all`.

- [X] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [X] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/nix-community/home-manager/blob/master/docs/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [X] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

  - [X] Added myself and the module files to `.github/CODEOWNERS`.

*Note: One has to set `programs.aerc.extraConfig.general.unsafe-accounts-conf = true;`, otherwise aerc will throw an error.*

Tested on NixOS 22.05, using smpt+tls+starttls and imap+tls, aswell as maildir via mbsync and sendmail via msmtp.

### CODE REVIEW

File: .github/CODEOWNERS [hunk 1]
```diff
@@ -480,3 +480,7 @@ Makefile                                              @thiagokokada
 /modules/services/plex-mpv-shim.nix                   @starcraft66
 
 /modules/services/sctd.nix                            @somasis
+
+/modules/programs/aerc.nix                            @lukasngl
+/modules/programs/aerc-accounts.nix                   @lukasngl
+/tests/modules/programs/aerc                          @lukasngl
```
Comment (sumnerevans) [inline .github/CODEOWNERS:486]: This file is alphabetized, so this should go before `/modules/programs/aria2.nix`

**Review by sumnerevans — COMMENTED** [review]
I haven't had time to review this in depth, but the overall structure looks fine.

**Comment by lukasngl** [comment]
Hi,

Thanks for the review.
I moved the entries in `.github/CODEOWNERS` and squashed the commits
to avoid clutter in the history.

**Comment by lukasngl** [comment]
Hi, I just fixed the failing workflow step `nix-build -A docs.jsonModuleMaintainers` on my part. I used `lib.maintainers` instead of `lib.hm.maintainers`.
(My branch is based on master and `nix-build -A docs.jsonModuleMaintainers` still fails locally, since @somasis is not in the maintainers set.)

**Comment by teto** [comment]
I've tried to test this but everytime I am greeted with the aerc assistant even when I enabled aerc for an account. Is that expected ?
I got the warning about 
```
extraConfig.general.unsafe-accounts-conf = true;
```
and I wonder if you could fully expand the warning to show "programs.aerc.extraConfig.general.unsafe-accounts-conf " and maybe link some page explaining what the setting entails.


NB: when in aerc configuration assistant, I cant find a way to exit it via mappings :/

**Comment by teto** [comment]
clearly this is a lot of effort. Let's merge and you can fix stuff later but it already looks good.

File: modules/programs/aerc-accounts.nix [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+      name = f k;
+      value = v;
+    }) attr));
+  addAccountName = name: k: "${k}:account=${name}";
+in {
+  type = mkOption {
+    type = types.attrsOf (types.submodule {
+      options.aerc = {
+        enable = mkEnableOption "aerc";
+        extraAccounts = mkOption {
```
Comment (rycee) [inline modules/programs/aerc-accounts.nix:18]: Please separate option definitions by empty line.

File: modules/programs/aerc-accounts.nix [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+with lib;
+
+let
+  mapAttrNames = f: attr:
+    with builtins;
+    listToAttrs (attrValues (mapAttrs (k: v: {
+      name = f k;
+      value = v;
+    }) attr));
+  addAccountName = name: k: "${k}:account=${name}";
```
Comment (rycee) [inline modules/programs/aerc-accounts.nix:12]: Please separate definitions by empty line to help readability.

File: modules/programs/aerc.nix [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+      let value = if builtins.isString v then v else toINI { global = v; };
+      in {
+        name = "aerc/stylesets/${k}";
+        value.text = joinCfg [ header value ];
+      });
+    mkTemplates = attrsets.mapAttrs' (k: v: {
+      name = "aerc/templates/${k}";
+      value.text = v;
+    });
+    accountsExtraAccounts = builtins.mapAttrs accounts.mkAccount aerc-accounts;
```
Comment (rycee) [inline modules/programs/aerc.nix:116]: I think `builtins.mapAttrs` could be simply `mapAttrs`.

**Comment by lukasngl** [comment]
> I've tried to test this but everytime I am greeted with the aerc assistant even when I enabled aerc for an account. Is that expected ?

Depends, if `programs.aerc.enable` is not set, enabling an account will not
create `accounts.conf` and thus aerc will start the assistant on launch.
Furthermore, aerc will create the files `aerc.conf` and `bindings.conf`,
if missing on launch.
This means that home-manager wont't create the `accounts.conf` if the
former files were already created and home-manager would need to overwrite them
(and display a warning).
Due to an issue fixed in [1], `aerc.conf` and `binds.conf` were always
created if an account (and the programs itself) was enabled.

Maybe you could checkout the fork [3], make sure `programs.aerc.enable`
is set to true and delete or backup (`home-manager switch -b <extension>`)
`aerc.conf` and `bindings.conf` if there is a conflict.

If the issue persists please provide further information,
so I can properly reproduce and investigate the issue.


> I got the warning about 
> ```
> extraConfig.general.unsafe-accounts-conf = true;
> ```
> and I wonder if you could fully expand the warning to show "programs.aerc.extraConfig.general.unsafe-accounts-conf " and maybe link some page explaining what the setting entails.

The warning was improved [2] and now explains the reasons and
implications and cites the relevant manpage section (aerc-config(5)).

> NB: when in aerc configuration assistant, I cant find a way to exit it via mappings :/ 

For me it was `Ctrl+q` to exit the assistant.

[1]: https://github.com/nix-community/home-manager/commit/3f8bed4105c237f2060493684014a16b19002082
[2]: https://github.com/nix-community/home-manager/commit/772574cc7870f40304c0f3b0e2df03b81cede7c6
[3]: https://github.com/lukasngl/home-manager/

File: modules/programs/aerc-accounts.nix (hunk 2 above)
Comment (lukasngl) [reply modules/programs/aerc-accounts.nix:18]: Resolved in #3150

**Pull request merged**